curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/implement-map-on-a-prototype.english.md

```diff
@@ -6,19 +6,21 @@ forumTopicId: 301230
 ---
 
 ## Description
+
 <section id='description'>
 As you have seen from applying <code>Array.prototype.map()</code>, or simply <code>map()</code> earlier, the <code>map</code> method returns an array of the same length as the one it was called on. It also doesn't alter the original array, as long as its callback function doesn't.
 In other words, <code>map</code> is a pure function, and its output depends solely on its inputs. Plus, it takes another function as its argument.
-It would teach us a lot about <code>map</code> to try to implement a version of it that behaves exactly like the <code>Array.prototype.map()</code> with a <code>for</code> loop or <code>Array.prototype.forEach()</code>.
-Note: A pure function is allowed to alter local variables defined within its scope, although, it's preferable to avoid that as well.
+You might learn a lot about the <code>map</code> method if you implement your own version of it. It is recommended you use a <code>for</code> loop or <code>Array.prototype.forEach()</code>.
 </section>
 
 ## Instructions
+
 <section id='instructions'>
-Write your own <code>Array.prototype.myMap()</code>, which should behave exactly like <code>Array.prototype.map()</code>. You may use a <code>for</code> loop or the <code>forEach</code> method.
+Write your own <code>Array.prototype.myMap()</code>, which should behave exactly like <code>Array.prototype.map()</code>. You should not use the built-in <code>map</code> method. The <code>Array</code> instance can be accessed in the <code>myMap</code> method using <code>this</code>.
```

Comment:
Would it be useful to have a reference/link to the lesson on `this` here? Asking campers to use `this` in the challenge without any context about it seems a bit sudden, so maybe a link to an earlier lesson would be handy (just in case someone wants a refresher).

Reply:
The earlier lesson where the user learns about `this` is [Make Code More Reusable with the this Keyword](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/object-oriented-programming/make-code-more-reusable-with-the-this-keyword).  That lesson's only explanation for using `this` is:
> 'this` refers to the object that the method is associated with: `duck`

Two reasons why the above statement would probably not help the user in the other lesson:
* It should probably state something closer to "`this` references the object (`duck`) that is currently calling the method"
* Even if it did state the above, most users would not understand at this point that an array is a special object in JavaScript

Reply:
I am going to create a separate issue around the [Make Code More Reusable with the this Keyword](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/object-oriented-programming/make-code-more-reusable-with-the-this-keyword), so someone can improve it first.  Then, I would not be opposed to linking to it from this lesson.

Reply:
I do agree asking learners to use `this` without covering it extensively in an earlier lesson would be unreasonable. This issue was actually opened due to the fact that most learners could not figure out how to access the `Array` instance using `this` when implementing `map` and `filter`. And that could be because `this` wasn't well explained in the previous lesson and in my opinion that  should be a different issue.

About adding a link to a lesson on `this`, I don't understand why that is really necessary. Essentially from what I have observed, almost every lesson on FCC either builds on or requires some knowledge of previous lessons because that is how the curriculum is structured. And to the best of my knowledge learners are encouraged to follow those lessons chronologically to derive maximum benefit. Why then must we keep linking to prior lessons that are necessary to complete/understand a particular lesson? 

If `this` is so important that learners need  reminding about by,
1-  providing a hint(which this PR is mostly about)
2-  providing a link to an earlier lesson which explains `this` 
 what about `for` loop or `forEach` method which learners are equally expected to know to complete the exercise?  Aren't they worth linking to? 

I think it is better to open an issue to make sure `this` is well explained in its own lesson so that by the time learners reach this exercise,  they know what it is and a simple reminder of how to access `Array` instance using `this` would suffice. Anyway I stand to be corrected.

Reply:
> Why then must we keep linking to prior lessons that are necessary to complete/understand a particular lesson?

Some concepts may have not been covered "in depth", so a link back to a previous lesson provides that quick reminder.

> I think it is better to open an issue to make sure this is well explained in its own lesson so that by the time learners reach this exercise,

There are no plans to add a new lesson on this in the existing curriculum.  We are stuck with modifying the existing lessons.

I think the thing to remember about the curriculum, is that it is not meant to be comprehensive.  We expect users to use the read, search, ask methodology.

@nibble0101 With the constraints I have stated above, can you explain how you would like to proceed on this PR?  I just want to make sure we are all on the same page, before you make any further changes on this PR.

Thank you for your patience.

Reply:
@RandellDawson Sorry for the late response. If a new lesson cannot be added, essentially what you are asking me to do is include snippets of information about `this` in [make code more reusable with the this keyword](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/object-oriented-programming/make-code-more-reusable-with-the-this-keyword) lesson primarily to enable the learner successfully implement [map](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/functional-programming/implement-map-on-a-prototype) and [filter](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/functional-programming/implement-the-filter-method-on-a-prototype) methods on a prototype. I have two problems with that.

1)  I can't think of concise way to explain `this` (at least the way I have understood `this`) in [make code more reusable with the this keyword](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/object-oriented-programming/make-code-more-reusable-with-the-this-keyword) lesson  without making the lesson more complicated and unnecessarily long.   
2)  I don't believe adding some information about `this` in [make code more reusable with the this keyword](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/object-oriented-programming/make-code-more-reusable-with-the-this-keyword) lesson primarily to enable the learner successfully implement [map](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/functional-programming/implement-map-on-a-prototype) and [filter](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/functional-programming/implement-the-filter-method-on-a-prototype) methods on a prototype can be construed as a solution to the core problem which is failure to cover `this` to a reasonable depth anywhere in the lesson series. To me what you are proposing is a workaround.

Given the constraints you stated above, I am proposing the following:
1) Open an issue that will fix [make code more reusable with the this keyword](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/object-oriented-programming/make-code-more-reusable-with-the-this-keyword) lesson because  like I stated above I don't know how to fix it without messing what is already there and even if I did, I don't believe it solves the core problem. Then after that I can modify this PR by adding a link. 
2) I close this PR and an issue is opened so that someone makes a PR  that fixes both lessons.